### PR TITLE
refactor arrayMin/arrayMax to have a generic specialization

### LINKED ISSUE #52756: arrayMin/arrayMax should have a generic specialization
**Use case**

```
milovidov-desktop :) SELECT arrayMin([3, 1, 2])

SELECT arrayMin([3, 1, 2])

Query id: 9bf45956-442f-47a6-af83-74f40361cb66

┌─arrayMin([3, 1, 2])─┐
│                   1 │
└─────────────────────┘

1 row in set. Elapsed: 0.001 sec. 

milovidov-desktop :) SELECT arrayMin([[3], [1], [2]])

SELECT arrayMin([[3], [1], [2]])

Query id: 1d8d574a-ce78-45a0-9bba-a5b80110dbc2


0 rows in set. Elapsed: 0.001 sec. 

Received exception from server (version 23.8.1):
Code: 43. DB::Exception: Received from localhost:9000. DB::Exception: array aggregation function cannot be performed on type Array(UInt8): While processing arrayMin([[3], [1], [2]]). (ILLEGAL_TYPE_OF_ARGUMENT)

milovidov-desktop :) SELECT arrayReduce('min', [[3], [1], [2]])

SELECT arrayReduce('min', [[3], [1], [2]])

Query id: f7053fe6-4639-4a2f-85bc-acc10592d8ab

┌─arrayReduce('min', [[3], [1], [2]])─┐
│ [1]                                 │
└─────────────────────────────────────┘

1 row in set. Elapsed: 0.001 sec.
```

**Describe the solution you'd like**

`arrayMin/Max` should be applicable to all data types that have `IColumn::compareAt`.

### PR DESCRIPTION
**Pull request by chloro-pn**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
arrayMin/Max can be applicable to all data types that are comparable

Closes #52756

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by chloro-pn** [comment]
There are currently some issues that need to be resolved (which may be added later)：
- The `nan_direction_hint` parameter (used for comparing Nullable type) is hard coded to 1 in the arrayMax/arrayMin function because I can't think of a good way to make the user specify its value.
-

**Comment by chloro-pn** [comment]
Good news, `01602_array_aggregation` pass:
捕获

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 9cc936697b3e2fd8d11a43ff4b29db5f65861ab5 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65455/9cc936697b3e2fd8d11a43ff4b29db5f65861ab5/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by chloro-pn** [comment]
failed on Integration test, here is the log:

---
```
E           Exception: Sanitizer assert found for instance =================================================================
E           ==1==ERROR: LeakSanitizer: detected memory leaks
E           
E           Direct leak of 240 byte(s) in 1 object(s) allocated from:
E               #0 0x55c82a0264cf in malloc (/usr/bin/clickhouse+0xa5114cf) (BuildId: 2d51419a5a8123a5847df2b8f1dd5bf5f7097896)
E               #1 0x55c8579e403e in CRYPTO_malloc build_docker/./contrib/openssl/crypto/mem.c:202:11
E               #2 0x55c8579e403e in CRYPTO_zalloc build_docker/./contrib/openssl/crypto/mem.c:222:11
E               #3 0x55c85792050a in ENGINE_new build_docker/./contrib/openssl/crypto/engine/eng_lib.c:36:16
E               #4 0x55c85285d554 in s2n_rand_init (/usr/bin/clickhouse+0x32d48554) (BuildId: 2d51419a5a8123a5847df2b8f1dd5bf5f7097896)
E               #5 0x55c8528586fa in s2n_init (/usr/bin/clickhouse+0x32d436fa) (BuildId: 2d51419a5a8123a5847df2b8f1dd5bf5f7097896)
E               #6 0x55c8527de562 in aws_tls_init_static_state (/usr/bin/clickhouse+0x32cc9562) (BuildId: 2d51419a5a8123a5847df2b8f1dd5bf5f7097896)
E               #7 0x55c8527dcb8c in aws_io_library_init (/usr/bin/clickhouse+0x32cc7b8c) (BuildId: 2d51419a5a8123a5847df2b8f1dd5bf5f7097896)
E               #8 0x55c85288f624 in aws_mqtt_library_init (/usr/bin/clickhouse+0x32d7a624) (BuildId: 2d51419a5a8123a5847df2b8f1dd5bf5f7097896)
E               #9 0x55c852864436 in Aws::Crt::ApiHandle::ApiHandle(aws_allocator*) (/usr/bin/clickhouse+0x32d4f436) (BuildId: 2d51419a5a8123a5847df2b8f1dd5bf5f7097896)
E               #10 0x55c8522be20d in Aws::InitializeCrt() (/usr/bin/clickhouse+0x327a920d) (BuildId: 2d51419a5a8123a5847df2b8f1dd5bf5f7097896)
E               #11 0x55c8522b8cf6 in Aws::InitAPI(Aws::SDKOptions const&) (/usr/bin/clickhouse+0x327a3cf6) (BuildId: 2d51419a5a8123a5847df2b8f1dd5bf5f7097896)
E               #12 0x55c84562997c in DB::S3::ClientFactory::ClientFactory() build_docker/./src/IO/S3/Client.cpp:879:5
E               #13 0x55c84562a0ba in DB::S3::ClientFactory::instance() build_docker/./src/IO/S3/Client.cpp:892:26
E               #14 0x55c84578d717 in DB::getClient(DB::S3::URI const&, DB::S3ObjectStorageSettings const&, std::__1::shared_ptr<DB::Context const>, bool) build_docker/./src/Disks/ObjectStorages/S3/diskSettings.cpp:87:60
E               #15 0x55c84549db8d in DB::StorageS3Configuration::createObjectStorage(std::__1::shared_ptr<DB::Context const>, bool) build_docker/./src/Storages/ObjectStorage/S3/Configuration.cpp:129:19
E               #16 0x55c84ac9c76b in DB::createStorageObjectStorage(DB::StorageFactory::Arguments const&, std::__1::shared_ptr<DB::StorageObjectStorage::Configuration>, std::__1::shared_ptr<DB::Context const>) build_docker/./src/Storages/ObjectStorage/registerStorageObjectStorage.cpp:61:24
```

It seems that it's because the sub module openssl has been updated
@alexey-milovidov

**Comment by antaljanosbenjamin** [comment]
Do you need the changes in the submodules (in folder `contrib`)? I think no, so could you please revert those changes?

**Comment by chloro-pn** [comment]
> Do you need the changes in the submodules (in folder `contrib`)? I think no, so could you please revert those changes?
> 
> Edit: that would probably also solve the issue swith LeakSan.

Actually, I just executed the command `git submodule update --init --recursive` to update the submodule, it looks strange, and I will try to fix it.

**Comment by antaljanosbenjamin** [comment]
If you need help, just let me know. I can fix it too, if you are running into issues.

**Comment by chloro-pn** [comment]
@antaljanosbenjamin hi bro , I have passed almost all the tests, but there are some that I cannot determine where the problem lies：
- **03040_dynamic_type_alters_2**  I passed the local test
- **test_ttl_move/test.py::test_alter_with_merge_work[mt_work]** I cannot determine where the problem lies through the log files

**Comment by chloro-pn** [comment]
@alexey-milovidov  hi, could you please merge my PR, or what else do I need to do

**Comment by chloro-pn** [comment]
@antaljanosbenjamin Okay, I  test **02435_rollback_cancelled_queries** locally and below are the exception message of the server, which may be helpful:

---
server version 24.7.1
```
2024.06.24 23:14:29.136246 [ 344755 ] {0d5327ea-1933-4a11-8302-6266249b892f} <Error> executeQuery: Code: 48. DB::Exception: Transactions are not supported: while starting a transaction with 'implicit_transaction'. (NOT_IMPLEMENTED) (version 24.7.1.1) (from 127.0.0.1:58058) (comment: 02435_rollback_cancelled_queries.sh) (in query: select throwIf(count() % 1000000 != 0 or count() = 0) from dedup_test), Stack trace (when copying this message, always include the lines below):

0. ./contrib/llvm-project/libcxx/include/exception:141: Poco::Exception::Exception(String const&, int) @ 0x00000000141be9d2
1. ./build/./src/Common/Exception.cpp:101: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000b4e8899
2. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000006708e0c
3. DB::Exception::Exception<>(int, FormatStringHelperImpl<>) @ 0x0000000006713b2b
4. ./build/./src/Interpreters/Context.cpp:0: DB::Context::checkTransactionsAreAllowed(bool) const @ 0x000000000fe64bec
5. ./contrib/llvm-project/libcxx/include/atomic:958: DB::InterpreterTransactionControlQuery::executeBegin(std::shared_ptr<DB::Context>) @ 0x0000000010ad386c
6. ./build/./src/Interpreters/InterpreterTransactionControlQuery.cpp:28: DB::InterpreterTransactionControlQuery::execute() @ 0x0000000010ad36c1
7. ./build/./src/Interpreters/executeQuery.cpp:0: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*)::$_0::operator()(std::shared_ptr<DB::Context> const&, DB::ASTTransactionControl::QueryType) const @ 0x0000000010acf941
8. ./build/./src/Interpreters/executeQuery.cpp:1161: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x0000000010ac793b
9. ./build/./src/Interpreters/executeQuery.cpp:1391: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x0000000010ac4b55
10. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:612: DB::TCPHandler::runImpl() @ 0x0000000011a94dab
11. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:593: DB::TCPHandler::run() @ 0x0000000011aac7d9
12. ./build/./base/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x0000000014262f27
13. ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:48: Poco::Net::TCPServerDispatcher::run() @ 0x00000000142633fa
14. ./build/./base/poco/Foundation/src/ThreadPool.cpp:219: Poco::PooledThread::run() @ 0x000000001420e5b2
15. ./base/poco/Foundation/include/Poco/SharedPtr.h:139: Poco::ThreadImpl::runnableEntry(void*) @ 0x000000001420c0a3
16. ? @ 0x00007f3ee1de1609
17. ? @ 0x00007f3ee1cfc353
```

**Comment by antaljanosbenjamin** [comment]
> I test 02435_rollback_cancelled_queries locally and below are the exception message of the server, which may be helpful:

So it is reproducible locally? If so, is it deterministic? If yes, than your changes has something to do with.

**Comment by chloro-pn** [comment]
> > I test 02435_rollback_cancelled_queries locally and below are the exception message of the server, which may be helpful:
> 
> So it is reproducible locally? If so, is it deterministic? If yes, than your changes has something to do with.

Yes, it can be reproduced locally and is deterministic, but the error message is different from the one given by CI (the error message tested locally is the part I posted in the reply above).
But this may be due to different versions of servers or different server startup configurations. (I am using embedded config locally)
I am currently deleting my PR locally and recompiling the code to see if the error messages are consistent.

**Comment by antaljanosbenjamin** [comment]
It didn't reproduce in other tests, so probably the issue is in your local environment.

**Comment by chloro-pn** [comment]
Yes, I reconfigured the startup parameters and the test passed.

**Comment by rschu1ze** [comment]
@chloro-pn This PR caused a [regression](https://github.com/ClickHouse/ClickHouse/issues/69620). If you find time, it would be great if you could take a look, thanks.

**Comment by chloro-pn** [comment]
> @chloro-pn This PR caused a [regression](https://github.com/ClickHouse/ClickHouse/issues/69620). If you find time, it would be great if you could take a look, thanks.

Okay, I will try to fix it

**Comment by alfonsonishikawa** [comment]
About generic specialization and comparable, maybe this tests should be added too?

select arrayMax(x -> x.1, [(1, 'a'), (0, 'b')])
result: 1

select arrayMax(x -> x.2, [(1, 'a'), (0, 'b')])
result: b

File: tests/queries/0_stateless/01602_array_aggregation.sql [hunk 1]
```diff
@@ -3,6 +3,12 @@ SELECT 'Array max ', (arrayMax(array(1,2,3,4,5,6)));
 SELECT 'Array sum ', (arraySum(array(1,2,3,4,5,6)));
 SELECT 'Array avg ', (arrayAvg(array(1,2,3,4,5,6)));
 
+SELECT 'Array min :';
```
Comment (Algunenano) [inline tests/queries/0_stateless/01602_array_aggregation.sql:6]: For the future: This is not ok. First it should have added a new file, and second it should have added way more tests. Adding just the simplest of cases lead to untested code, and as it happened, to multiple bugs being introduced.

**Comment by antaljanosbenjamin** [comment]
For some reason I missed these issues. Sorry for causing troubles, I will try to be more careful in the future.

**Pull request merged**